sdk/core/core-http/src/policies/bearerTokenAuthenticationPolicy.ts

```diff
@@ -10,107 +10,210 @@ import {
- * @param scopes - The scopes for which the bearer token applies.
+ * @param tryGetAccessToken - a function that produces a promise of an access
+ * token that may fail by returning null
+ * @param retryInterval - the time (in milliseconds) to wait between retry
+ * attempts
+ * @returns - a promise that, if it resolves, will resolve with an access token
  */
-export function bearerTokenAuthenticationPolicy(
-  credential: TokenCredential,
-  scopes: string | string[]
-): RequestPolicyFactory {
-  const tokenCache: AccessTokenCache = new ExpiringAccessTokenCache();
-  const tokenRefresher = new AccessTokenRefresher(
-    credential,
-    scopes,
-    timeBetweenRefreshAttemptsInMs
-  );
+async function beginRefresh(
+  tryGetAccessToken: () => Promise<AccessToken | null>,
+  retryInterval: number
+): Promise<AccessToken> {
+  let token: AccessToken | null = await tryGetAccessToken();
 
-  return {
-    create: (nextPolicy: RequestPolicy, options: RequestPolicyOptions) => {
-      return new BearerTokenAuthenticationPolicy(nextPolicy, options, tokenCache, tokenRefresher);
-    }
-  };
+  while (token === null) {
+    await delay(retryInterval);
+
+    token = await tryGetAccessToken();
+  }
+
+  return token;
 }
 
 /**
+ * Creates a token cycler from a credential, scopes, and optional settings.
+ *
+ * A token cycler represents a way to reliably retrieve a valid access token
+ * from a TokenCredential. It will handle initializing the token, refreshing it
+ * when it nears expiration, and synchronizes refresh attempts to avoid
+ * concurrency hazards.
  *
- * Provides a RequestPolicy that can request a token from a TokenCredential
- * implementation and then apply it to the Authorization header of a request
- * as a Bearer token.
+ * @param credential - the underlying TokenCredential that provides the access
+ * token
+ * @param scopes - the scopes to request authorization for
+ * @param _options - optionally override default settings for the cycler
  *
+ * @returns - a function that reliably produces a valid access token
  */
-export class BearerTokenAuthenticationPolicy extends BaseRequestPolicy {
-  /**
-   * Creates a new BearerTokenAuthenticationPolicy object.
-   *
-   * @param nextPolicy - The next RequestPolicy in the request pipeline.
-   * @param options - Options for this RequestPolicy.
-   * @param credential - The TokenCredential implementation that can supply the bearer token.
-   * @param scopes - The scopes for which the bearer token applies.
-   * @param tokenCache - The cache for the most recent AccessToken returned from the TokenCredential.
-   */
-  constructor(
-    nextPolicy: RequestPolicy,
-    options: RequestPolicyOptions,
-    private tokenCache: AccessTokenCache,
-    private tokenRefresher: AccessTokenRefresher
-  ) {
-    super(nextPolicy, options);
-  }
+function createTokenCycler(
+  credential: TokenCredential,
+  scopes: string | string[],
+  _options?: Partial<TokenCyclerOptions>
```

Comment:
I get that you wanted to use the nice name `options` for the final shape below, but since TS uses underscore prefix to mean "this parameter is never used in the method body" I was confused at first

Reply:
This is just something I missed. Should be `tokenCyclerOptions`.